File: CCPlayer.UWP.Pro/StateTriggers/BooleanDataTrigger.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.UI.Xaml;

namespace CCPlayer.UWP.StateTriggers
{
    
    public class BooleanDataTrigger : StateTriggerBase
    {
        #region DataValue
        public static Boolean GetDataValue(DependencyObject obj)
        {
            return (Boolean)obj.GetValue(DataValueProperty);
        }

        public static void SetDataValue(DependencyObject obj, Boolean value)
        {
            obj.SetValue(DataValueProperty, value);
        }

        public static readonly DependencyProperty DataValueProperty =
            DependencyProperty.RegisterAttached("DataValue", typeof(Boolean),
                typeof(BooleanDataTrigger), new PropertyMetadata(false, DataValueChanged));

        private static void DataValueChanged(DependencyObject target, DependencyPropertyChangedEventArgs e)
        {
            Boolean triggerValue = (Boolean)target.GetValue(BooleanDataTrigger.TriggerValueProperty);
            TriggerStateCheck(target, (Boolean)e.NewValue, triggerValue);
        }
        #endregion

        #region TriggerValue

        public static Boolean GetTriggerValue(DependencyObject obj)
        {
            return (Boolean)obj.GetValue(TriggerValueProperty);
        }

        public static void SetTriggerValue(DependencyObject obj, Boolean value)
        {
            obj.SetValue(TriggerValueProperty, value);
        }

        public static readonly DependencyProperty TriggerValueProperty =
                DependencyProperty.RegisterAttached("TriggerValue", typeof(Boolean),
                    typeof(BooleanDataTrigger), new PropertyMetadata(false, TriggerValueChanged));

        private static void TriggerValueChanged(DependencyObject target, DependencyPropertyChangedEventArgs e)
        {
            Boolean dataValue = (Boolean)target.GetValue(BooleanDataTrigger.DataValueProperty);
            TriggerStateCheck(target, dataValue, (Boolean)e.NewValue);
        }
        #endregion

        private static void TriggerStateCheck(DependencyObject target, Boolean dataValue, Boolean triggerValue)
        {
            BooleanDataTrigger trigger = target as BooleanDataTrigger;
            if (trigger == null) return;
            trigger.SetActive(triggerValue == dataValue);
        }
    }
}

File: CCPlayer.UWP.Xaml.Control/StateTriggers/BooleanDataTrigger.cpp
#include "pch.h"
#include "BooleanDataTrigger.h"

using namespace CCPlayer::UWP::Xaml::StateTriggers;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Interop;

DependencyProperty^ BooleanDataTrigger::_DataValueProperty = 
	DependencyProperty::RegisterAttached(
		"DataValue", 
		TypeName(bool::typeid), 
		TypeName(BooleanDataTrigger::typeid),
		ref new PropertyMetadata(
			true, 
			ref new PropertyChangedCallback(&BooleanDataTrigger::OnDataValueChanged)
		)
	);

DependencyProperty^ BooleanDataTrigger::_TriggerValueProperty = 
	DependencyProperty::RegisterAttached(
		"TriggerValue", 
		TypeName(bool::typeid), 
		TypeName(BooleanDataTrigger::typeid),
		ref new PropertyMetadata(
			true, 
			ref new PropertyChangedCallback(&BooleanDataTrigger::OnTriggerValueChanged)
		)
	);

bool BooleanDataTrigger::GetDataValue(DependencyObject^ obj)
{
	return (bool)obj->GetValue(DataValueProperty);
}

void BooleanDataTrigger::SetDataValue(DependencyObject^ obj, bool value)
{
	obj->SetValue(DataValueProperty, value);
}

bool BooleanDataTrigger::GetTriggerValue(DependencyObject^ obj)
{
	return (bool)obj->GetValue(TriggerValueProperty);
}

void BooleanDataTrigger::SetTriggerValue(DependencyObject^ obj, bool value)
{
	obj->SetValue(TriggerValueProperty, value);
}

void BooleanDataTrigger::TriggerStateCheck(DependencyObject^ target, bool dataValue, bool triggerValue)
{
	BooleanDataTrigger^ trigger = dynamic_cast<BooleanDataTrigger^>(target);
	if (trigger == nullptr) return;
	trigger->SetActive(triggerValue == dataValue);
}

void BooleanDataTrigger::OnDataValueChanged(DependencyObject^ target, DependencyPropertyChangedEventArgs^ e)
{
	bool triggerValue = (bool)target->GetValue(BooleanDataTrigger::TriggerValueProperty);
	TriggerStateCheck(target, (bool)e->NewValue, triggerValue);
}

void BooleanDataTrigger::OnTriggerValueChanged(DependencyObject^ target, DependencyPropertyChangedEventArgs^ e)
{
	bool dataValue = (bool)target->GetValue(BooleanDataTrigger::DataValueProperty);
	TriggerStateCheck(target, dataValue, (bool)e->NewValue);
}

